## 1. Perkenalan

- Nama  : I Made Dhanan Pradipta
- Batch : 10

## 2. Problem

    Kamu adalah seorang data scientist yang bekerja di salah satu perusahaan manufaktur baterai Lithium ion. Tim produksi meminta kamu untuk mengecek apakah baterai yang mereka buat dalam kondisi bagus atau tidak. Sebelum melakukan pengecekan lebih lanjut, kamu diminta untuk menghitung hambatan/resistansi dan Daya setiap detik:

[R=\frac{V}{I}]

[P=VI]

    R adalah Kapasitas dalam satuan Ohm
    P adalah Daya dalam Watt
    I adalah kuat arus dalam satuan Ampere
    V adalah tegangan dalam satuan Volt


##### 1. Import Library

In [1]:
import pandas as pd
import numpy as np

##### 2. Data Loading

In [2]:
df = pd.read_excel("dataset/NMC1-9.xlsx")

In [3]:
df.head()

time(s)    V_m  Current (A)  Cycle       Cap
0        0  4.098        0.997      1  2.840817
1        1  4.096        0.998      1  2.840817
2        2  4.094        0.999      1  2.840817
3        3  4.093        0.998      1  2.840817
4        4  4.092        0.999      1  2.840817

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


##### 3. Answer

##### a. Buatlah dua fungsi eksplisit masing-masing memuat perhitungan hambatan (R) dan daya (P) yang nantinya akan digunakan untuk perhitungan

In [5]:
#mencari hambatan (R)
def hambatan(tegangan, arus):
    return tegangan / arus

In [6]:
#mencari daya (P)
def daya(tegangan, arus):
    return tegangan * arus

b. Buat kolom baru dengan nama kolom R/P berisikan hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, Kapasitas <2.5 hitung nilai hambatan. (Hint: Gunakan for loop dan if statement untuk mengerjakan perintah nomor b, kamu bisa menggunakan list untuk menyimpan perhitungan sementara, setelahnya dapat dimasukkan ke kolom R/P).

##### Versi 1 (kolom R/P tergabung dengan main dataframe (df))

In [12]:
#looping utk mencari hambatan dengan kriteria >=2.5
lst_hambatan = []
lst_daya = []
for i in range(len(df['V_m'])):
    R = hambatan(df['V_m'][i], df['Current (A)'][i]) #masukan function hambatan
    P = daya(df['V_m'][i], df['Current (A)'][i]) #masukan function daya
    lst_hambatan.append(R)
    lst_daya.append(P)

##### Versi 2 kolom R/P tidak tergabung dengan df

In [9]:
lst = []
for i in range(len(df['V_m'])):
    R = hambatan(df['V_m'][i], df['Current (A)'][i]) #masukan function hambatan
    P = daya(df['V_m'][i], df['Current (A)'][i]) #masukan daya
    if df['Cap'][i] < 2.5: #jika kondisi terpenuhi masukan nilai ke list kosong 
        lst.append(R)
    else:
        lst.append(P)

df['R/P'] = lst

##### c. Hitunglah rata-rata nilai kapasitas untuk arus = 1 Ampere dan cycle ganjil.

In [10]:
df.head()

time(s)    V_m  Current (A)  Cycle       Cap         R       R/P
0        0  4.098        0.997      1  2.840817  4.110331  4.085706
1        1  4.096        0.998      1  2.840817  4.104208  4.087808
2        2  4.094        0.999      1  2.840817  4.098098  4.089906
3        3  4.093        0.998      1  2.840817  4.101202  4.084814
4        4  4.092        0.999      1  2.840817  4.096096  4.087908

In [11]:
#menghitung nilai rata2 dari kapasitas (R) dengan kondisi Current (A) = 1 dan Cycle berupa angka ganjil

df['Cap'][(df['Current (A)'] == 1) & (df['Cycle'] % 2 == 1)].mean()

2.8064845553927484

rata-rata dari kapasitas (R) dengan kondisi Kuat Arus (A) = 1 dan juga Cycle bernilai ganjil yaitu 2.806 Ohm 